### ISSUE TITLE
Allow widening the foreign key compatibility criteria

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#24007]
ForeignKey.AreCompatible requires that the foreign and principal keys' CLR properties [correspond exactly](https://github.com/dotnet/efcore/blob/main/src/EFCore/Metadata/Internal/ForeignKey.cs#L1467). While SQL Server is indeed very strict about that, other database support wider foreign key types pointing to narrower principal keys.

On SQL Server:

```sql
CREATE TABLE posts (id INT PRIMARY KEY);
CREATE TABLE polls (post_id BIGINT REFERENCES posts(id));
```

... errors with `Column 'posts.id' is not the same data type as referencing column 'polls.post_id' in foreign key 'FK__polls__post_id__5165187F'.`.

However, the same DDL works fine on PG and Sqlite. We could allow some sort of provider hook that determines whether the properties types are compatible.

Originally flagged by @jessicah in https://github.com/npgsql/efcore.pg/issues/1515

**Comment by roji** [comment]
Duplicate of #19793

**Comment by roji** [comment]
Although it is possible to make EF Core support different types for principal and foreign keys, that's likely quite difficult to do. On the other hand, as specified in #19793, it's possible to simply specify the different store type in the model.

@jessicah although this wouldn't unblock the scaffolding scenario, try to simply define your foreign key as an INT in your model (instead of BIGINT) and everything should work.
